sip/sip-006-runtime-cost-assessment.md

```diff
@@ -0,0 +1,570 @@
+block, miners are free to make their own choices based on transaction
+fees, however, blocks may not exceed the _block limit_ (if they do so,
+the leader forfeits all rewards from the block).
+
+# Static versus Dynamic Cost Assessment
+
+Tracking the execution cost of a contract may be done either dynamically
+or statically. Dynamic cost assessment involves tracking, at the VM level,
+the various metrics as a contract is executed. Static cost assessment is
+performed via analysis of the contract source code, and is inherently
+a more pessimistic accounting of the execution cost: list operations
+are charged according to the _maximum_ size of the list (per the type
+annotations and inferences from the source code) and branching statements
+are charged according to the _maximum_ branch cost (per metric tracked, i.e.,
+if one branch performs 1 write and has a runtime cost of 1, and another
+branch performs 0 writes and has a runtime cost of 2, the whole statement
+will be assessed as having a maximum of 1 write and runtime cost of 2).
+
+# Costs of Common Operations
+
+### Variable Lookup
+
+Looking up variables in Clarity incurs a non-constant cost -- the stack
+depth _and_ the length of the variable name affect this cost.
+
+The stack depth affects the lookup cost because the variable must be
+checked for in each context on the stack. The variable name affects
+the cost because variables are stored in a HashMap, and determining
+the key associated with a variable name requires hashing that name (a
+linear operation).
+
+Cost Function:
+
+```
+(a*X+b)*Y+c
+```
+
+where a, b, and c are constants,
+X := variable name length
+Y := stack depth
+
+### Function Lookup
+
+Looking up a function in Clarity incurs a non-constant cost, which
+depends on the length of the function name (for the same reason variable
+lookup depends on name length). However, because functions may only
+be defined in the top-level contract context, stack depth does not
+affect function lookup.
+
+Cost Function:
+
+```
+(a*X+b)
+```
+
+where a and b are constants,
+X := function name length
+
+### Name Binding
+
+The cost of binding a name in Clarity -- in either a local or the contract
+context is _linear_ in the length of the name:
+
+```
+binding_cost(X) = (a*X+b)
+```
+
+where a, b are constants,
+X := the length of the bound name
+
+### Function Application
+
+Function application in Clarity incurs a cost in addition to the
+cost of executing the function's body. This cost is the cost of
+binding the arguments to their passed values, and the cost of
+ensuring that those arguments are of the correct type. Type checks
+and argument binding are _linear_ in the size of the arguments, and
+also incur a linear cost in hashing the argument names.
+
+The cost of applying a function is:
+
+
+```
+(a*X+b) + sum(binding_cost(Y), ARGS) + costEval(body)
+```
+
+where a and b are constants,
+X := the cumulative size of the argument types,
```

Comment:
I think `(a*X + b)` refers to the function lookup.  `Y` should be defined as the length of the _i_th argument, I think?

Reply:
This section might be confusingly described. Also, perhaps confusingly, I don't defined new constant names for each function in this SIP -- rather each constant is scoped to its section.

The function lookup cost isn't assessed in this equation -- a function lookup is constant (since functions are only defined in the top-level scope of a contract -- so a single hashmap lookup) or in the case of a contract-call, includes the cost of loading a contract.

(a*X+b) is the cost of type checking supplied arguments -- type checks are linear with respect to the size of their describes objects (e.g., type-checking a `(list int 5)` is `(a*5+b)`). The process for a function application is:

1. Type check arguments.
2. Bind the arguments to a new local scope (this used to be described as variable in argument names, but each binding is constant because variable names are bounded).
3. Evaluate the function body.